### ISSUE TITLE
Create a generic "Table" class

### ISSUE THREAD

**Issue by inbalzukerman** [TechnionYP5777/SmartCity-Elderly-Support#63]
The table will be used to save information received from the sensors in the system database.

**Comment by inbalzukerman** [comment]
Currently the SensorInformationDatabase class has an ArrayList of tuples, where each tuple saves a "record" sent from the sensor. Our idea on the meeting yesterday was to change the implementation and use a generic table (this is why I've opened this issue)

After discussing with the sensor team, it was decided that the system will recieve the information from the sensors as a map of <row names, values>

My question- why not to change the SensorInformationDatabase to save an ArrayList of maps? (instead of ArrayList of tuples)

I think that creating a class for the table itself is unnecessary since it will only be used in the SensorInformationDatabase  class anyway and this class is responsible to retrieve the data for the applications 

If you agree please check your name:

- [ ] @EliaTraore 
- [ ] @SharonKL 
- [ ] @ylevv 
- [ ] @roy-shchory 
- [ ] @RonGatenio

**Comment by RonGatenio** [comment]
I think it will be better to work with a Table object, and the SensorInformationDatabase could have a list of Table objects. 

Some PROS for making a Table object:
- We could change the Table implementation easily if we want.
- An application could ask the system to return all of the information about a sensor. The system could return a Table (and not a list of stuff). This will be easier for the developers.

#just_a_thought
@inbalzukerman

**Comment by inbalzukerman** [comment]
But-
* This implementation is open for changes that can be easily done

* an application can receive an instance of SensorsInformationDB

**Comment by RonGatenio** [comment]
So SensorInformationDatabase is one table?

**Comment by inbalzukerman** [comment]
In a different class - Sensors Handler
I agree that the name is confusing, @EliaTraore mentioned it and I will change it

**Comment by RonGatenio** [comment]
I thought SensorsHandler handles the sensors 😆 (like the communication, registration and stuff.... the server should be implemented in this class)
And I thought SensorInformationDatabase is only the database

**Comment by EliaTraore** [comment]
this discussion is a bit confusing... :sweat_smile: 
if i'm understanding correctly, @inbalzukerman you want to:
(1) implement the database inside SensorInformationDatabase [SID] as an arraylist of maps
(2) send an instance of SID to the application
am I correct?

and you said all the tables will be stored in SensorHandler? it thought it was suppose to be in DatabaseHandler?

**Comment by inbalzukerman** [comment]
Some clarifications
The database of information from each sensor will be saved in an instance of SID (the name will be changed)

The application will receive- last update as a map or last k updates as an instance of SID with desirable k last updated.  

In the SID class implementation of methods such as - get last update,  etc. will be found

There will be a class responsible to the database management and all the updates from all the sensors will be saved there. In this class for each sensor an instance of SID will be saved and updated when needed


~names of classes will be changed for more clearity of their use~

**Comment by RonGatenio** [comment]
So if I understand you correctly:
SID == Table
SensorHandler == SensorInformationDatabase 
(current names == my names)

**Comment by inbalzukerman** [comment]
Yes, pretty much :disappointed:

**Comment by SharonKL** [comment]
I agree with Shushi's comment - The `SensorHandler`, which IMO should be changed to `SensorsHandler` is responsible for communication with the sensors.  The `SensorInformationDatabase` is responsible for storage of information.

The system is simply a class that will contain the `SensorsHandler` and the `AppsHandler`. Handlers allow communication, while the database allows storing and passing information.

**Comment by EliaTraore** [comment]
In that case, I support Ron's opinion - an abstraction could be better here.
Sending a list of maps reveals and creates high coupling with this implementation, and as we mentioned in #52 (that once upon a time described the SID) - we might want to change the implementation all together to a more robust DB. 
If that happens we will have to emulate sending lists of maps or refactor the whole code (which isn't as easy as it sounds. usually - refactor + generics ~= problems, even if you don't consider all the method invocations that are now invalid) 
This might seem like a next semester problem, but it influences our development at the moment as well (this change will result in refactor in the existing code).

And a question - in its current version, SID contains additional information like the sensorId, the table maxCapacity etc. will that remain part of the implementation? seems redundant to send that to the app everytime....

**Comment by SharonKL** [comment]
As for the `Table` class - I think it would be a better idea create the generic class, instead of using a Map of Lists in the `SensorInformationDatabase` considering this is not the only database we have to maintain. Even in this early stage we want to save user's information, in the upcoming semester we might add additional types of databases, that might benefit from a generic table type.

The `SensorInformationDatabase` might act as another layer on top of the `Table` class, defining more 'specific' functions for working with sensors databases. It also might be redundant and we might find it better to just have a `Table` object for each sensor the system maintains.

**Comment by inbalzukerman** [comment]
OK I think there is a misunderstanding-
The SID will be some kind of a table. It will save the information for each sensor but -
Saving for each sensor it's SID instance will be the responsibility of a different class!

The application will not receive an instance of the currently implemented SID, but - after changes in the SID class, and it will not contain the sensorID etc. No one will send the application a list of maps, this is not the idea... The idea is to make changes so the SID class will be transformed to a generic datatype simulating a table; a different class will save for each sensor it's ID and "table" and will implement the methods for the application's side to use

The modified SID class will not be specifically intended for saving information from the sensors, and, by my opinion, will be qualified to save any other data we'll might need.

What I tried to ask was if you all agree to - change and the SID class to be our shinning new "table" (appropriate modifications will be made by me) and that the `sensorDatabase` and the `sesnorHandler` will do the rest

I am  sorry if my previous explanations were not clear enough 😟

**Comment by EliaTraore** [comment]
So we were talking about the same thing all along? smh #existentialLoneliness
why not leave the SID as the specific table and have it implement the interface of a generic table (or inherit/have instance of a generic table)

**Comment by inbalzukerman** [comment]
SID will be the generic table 😕

**Comment by SharonKL** [comment]
So you want to make `SID` our generic table implementation, renaming it something like `Table`, while having other classes extend or use this generic table for specific tasks (such as saving sensors data, user registration data, etc...)? If so, I agree with you, I think we all got confused by the original name of `SID`.

As I see it, we would have something similar to the following structure:

* A generic `Table` class will be created.
* More specific table classes will be created that extend this generic table implementation. For example, `SensorTable` will contain an instance of this `Table` class and other functions to deal with sensors' data.
* The `SensorsHandler` class will maintain such a table **for each** sensor, while allowing the `AppsHandler` to receive information from these tables and send it to the applications.

**Comment by inbalzukerman** [comment]
@SharonKL you always explain thing so well 😍 
You understood me fully this time 😄 This is the structure i was trying (not so successfully) to explain 

As I said, the names are definitely confusing and i am changing them.

**Comment by EliaTraore** [comment]
@SharonKL , 

> The SensorsHandler class will maintain such a table for each sensor, while allowing the AppsHandler to receive information from these tables and send it to the applications.
doesn't that tie to separate jobs? managing the database and managing the sensors are 2 different issues...

**Comment by yossigil** [comment]
Good leadership Inbal!

**Comment by SharonKL** [comment]
@EliaTraore 

> doesn't that tie 2 separate jobs? managing the database and managing the sensors are 2 different issues...

Of course, the database will be managed by some class, extending the new `Table` class. The `SensorsHandler` (which is in-charge of handling communication with sensors) will use the database(s) to store new content. The `AppsHandler` will also use this database, but to receive the content.

**Comment by EliaTraore** [comment]
@inbalzukerman, please add a get method :sweat_smile: currently you can only get a table from the table, the apps will probably need something like 
```java
Object get(int row, String col);
```
and if the could be consts like `NEWEST`, `OLDEST` to have access to the newest and oldest rows of data more readable that would be great too :smile:

**Comment by inbalzukerman** [comment]
@EliaTraore - `Table` is not responsible to communicate with the applications. As @SharonKL explained very well, `SensorTable` will have an instace of `Table` and will implement the methods relevant to handling the sensors' data.

**Comment by SharonKL** [comment]
@inbalzukerman I think (from my understanding) that @EliaTraore would like to have more methods to the `Table` class for easier information extraction.

Both `SensorsHandler` and `AppsHandler` need to communicate with `Table` objects, the former will add content, while the latter will poll content from the table, but both need to communicate with it. We can of course expose different interfaces - the `SensorsHandler` will receive an interface that allows addition of new content to the database, while the `AppsHandler` will receive an interface that allows polling. But imo this is too much, especially when we need a working prototype soon.

tl;dr - both `SensorsHandler` and `AppsHandler` communicate with the table, but for different purposes. @EliaTraore asked for more functionality from the `Table` class.

**Comment by EliaTraore** [comment]
Someone somewhere will need to access the actual info in the table. currently you can only add entries, but you can never get them out (as a map or as a single object, which is preferable imo).

**Comment by EliaTraore** [comment]
The only getter methods in the table which allow getting info from the table are:
```java
// @return the last entry added to the table
public Map<L, R> getLastEntry();
// @return the value co-respond with the key of the last entry which is required
public R getLastEntryAtCol( final L key);
```
This means that in order to access info in any entry that isn't the last, you must call 
`public Table<L,R> receiveKLastEntries(final int numOfEntries); `
and get a smaller table first.
can a `public Object get(Object row, Object col);` be added?

**Comment by inbalzukerman** [comment]
@EliaTraore , you're right, my bad. Will be added 👍 😅
